### ISSUE TITLE
Example given an empty SVG

### ISSUE THREAD

**Issue by joepavitt** [gorenje/node-red-flowviewer-js#1]
Hi, we're interested in utilising this library in order to enhance https://flows.nodered.org/

Having played around with the library, am struggling to work out the base requirements for the `<svg></svg>` element for this to work. It would be great to know, given an empty SVG, what code do I need to run here?

My understanding was something like this:

```js
const id = "46186ccb85f93a93"
const svg = $('.flow-preview svg')
renderFlow(id, flow, svg, { 
   gridlines: true, 
   images: true, 
   linklines: true 
});
```

Where `flow` would be the full `flow.json`.

But looking at the errors I'm seeing, and the source code itself, seems like it can't just be an empty SVG at the moment?

**Comment by gorenje** [comment]
Hi,

Indeed, the SVG is assumed to contain a number of groups into which "stuff" is rendered.

Because the code generates the SVG into the `<code>` block in which the SVG will be, this isn't so obvious from `renderFlow` function. This [bit of code](https://github.com/gorenje/node-red-flowviewer-js/blob/1494f5996256c7ea8e8d60fd83194423705d67ac/src/replaceCodeBlocksWithNodeRedFlowImages.js#L18-L31) generates the SVG and also all the groups it requires.

If those groups (e.g. `flowGridLines`) aren't present then not much will happen - as you've seen ;)

**Comment by joepavitt** [comment]
Thanks @gorenje 

Using the following `flow.json`:

```json
[{"id":"46186ccb85f93a93","type":"tab","label":"Flow 2","disabled":false,"info":"","env":[]},{"id":"1eabd31a6d56d2e1","type":"inject","z":"46186ccb85f93a93","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":160,"y":100,"wires":[["6563b868847ca78a"]]},{"id":"6563b868847ca78a","type":"debug","z":"46186ccb85f93a93","name":"debug 1","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":320,"y":120,"wires":[]}]
```

And running:

```html
<div class="flowviewer">
    <svg>
        <g class="flowGridlines"></g>
        <g class="containerGroup"></g>
        <g class="flowGroups"></g>
        <g class="flowWires"></g>
        <g class="flowNodes"></g>
    </svg>
</div>
```

```js 
renderFlow("46186ccb85f93a93", flow, $('.flowviewer svg'), { 
    gridlines: true, 
    images: true, 
    linklines: true 
});
```

Results in:

<img width="632" alt="Screenshot 2023-09-01 at 10 14 50" src="https://github.com/gorenje/node-red-flowviewer-js/assets/99246719/0e3077c2-bb0b-48a3-9ca8-3045487a6442">

Note, no labelling, ~no grid~, and no interaction/moving of the canvas either.

Any guidance on what I've missed is most appreciated.

update: missing grid was a problem in my CSS imports, that's now resolved, other problems still exist.

**Comment by joepavitt** [comment]
I also tried:

```html
<div class="flowviewer-nosvg"></div>
```

```js
renderSvgWithOptions($('.flowviewer-nosvg'), "46186ccb85f93a93", flow, { 
   gridlines: true, 
   images: true, 
   linklines: true 
});
```

With the same rendered outcome

**Comment by gorenje** [comment]
i think it should be this:

```
<div class="flowviewer">
    <svg>
        <g class="flowGridlines"></g>
        <g class="containerGroup">       
           <g class="flowGroups"></g>
           <g class="flowWires"></g>
           <g class="flowNodes"></g>
      </g>
    </svg>
</div>
```
i.e. the groups, wires and nodes are in the [containerGroup-group](https://github.com/gorenje/node-red-flowviewer-js/blob/1494f5996256c7ea8e8d60fd83194423705d67ac/src/replaceCodeBlocksWithNodeRedFlowImages.js#L29-L31)!

**Comment by gorenje** [comment]
Hm, I was thinking about this and thought maybe it was the layering of the groups but the order seems to be ok. Also the text doesn't have a separate layer. I rendered the [flow using flowhub](https://flowhub.org/f/46186ccb85f93a93) and that also worked. About the only think I could suggest is that css is missing?

**Comment by joepavitt** [comment]
Yeah, I checked the rendered DOM and the text is never added. Haven't quite been able to track the exact problem yet.

**Comment by gorenje** [comment]
try adding ```"labels": true,``` to the options list ... options are overwritten and labels are probably assumed to be false then...

Edit: [this line](https://github.com/gorenje/node-red-flowviewer-js/blob/2aad141161f753150f7c3c10eae15beb37700336/src/renderFlow.js#L353) is preventing the labels from being rendered, it's the missing option

**Comment by joepavitt** [comment]
> try adding "labels": true, to the options list

We have labels! Just need the interaction/dragging now? Tried `zoom: true`, but not happening.

**Comment by gorenje** [comment]
I gave up trying to get zoom working and instead began using d3:

```
<body>

<script>
      window.addEventListener('load', function () {
        setTimeout( function(){
          var svgs = d3.selectAll(".svg-container-noderedjson svg");
          svgs.each(function() {
            var svg = d3.select(this);
            
            svg.html("<g>" + svg.html() + "</g>");
            var inner = svg.select("g");
            var zoom = d3.zoom().on("zoom", function(event) {
              inner.attr("transform", event.transform);
            });
            svg.call(zoom);
          });
        }, 500);
      });
</script>

</body>
```

after this:
```
<head>

  <script src="https://unpkg.com/d3/dist/d3.min.js" crossorigin="anonymous" referrerpolicy="no-referrer"></script>

</head>
```

see the bottom of the [blog post](https://blog.openmindmap.org/blog/backticks-in-markdown-and-node-red)

**Comment by gorenje** [comment]
closed with #2
